### DISCUSSION TITLE
[WASM][WindowsAppSDK][.NET6] Launcher.LaunchUriAsync not working

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#7773] [Q&A]
My custom app manages documents which I wanted to download and open similar to how it works on Windows desktop.

I'm trying to create and launch a document file following [directions here](https://platform.uno/docs/articles/features/windows-system.html#launchuriasync):
```
    CreateAndPopulateFileWithContent("/temp/text.txt");   // pseudo code
    await global::Windows.System.Launcher.LaunchUriAsync(new Uri("file:///temp/text.txt"));
```
However, it doesn't work: no browser tab nor window is opened. No exception is thrown, but the browser console shows error:
```
    Not allowed to load local resource: "file:///temp/text.txt"
```
Any idea what I'm doing wrong?

**Comment by MartinZikmund** [comment]
Hey @DierkDroth, unfortunately this kind of interaction is not possible, as the file is located on local drive and browser rules prevent sites from opening "file:" URIs directly for security reasons. Browser itself can open a `file:` URI when double-clicking in file explorer for example, but it is not possible to open it from JS (and by extension from WASM)

**Reply by DierkDroth** [reply]
@MartinZikmund thanks for your feedback.

Understand. Ok, is there any other option to achieve what I wanted to do (in UNO WASM = similar to what I'm doing in WinUI3):
- I have a byte[] array of data which basically is the content of document - say a PDF
- I needed to show/open that document 

How could that be done?

**Reply by MartinZikmund** [reply]
I am afraid such a feature is not yet available out of the box. Maybe a solution would be to trigger "download" of the byte data, would that suit your scenario?

**Comment by DierkDroth** [comment]
Thanks @MartinZikmund 

> Maybe a solution would be to trigger "download" of the byte data

Not sure what you mean by that. Could you elaborate?

**Reply by DierkDroth** [reply]
@MartinZikmund the best I could come up with is this:

```
// https://blog.jayway.com/2017/07/13/open-pdf-downloaded-api-javascript/
// https://ionicframework.com/blog/converting-a-base64-string-to-a-blob-in-javascript/
public static async Task OpenAsDocument(string filename, byte[] data)
{
	await Uno.Foundation.WebAssemblyRuntime.InvokeAsync($"(async () => {{ "
		+ $"const data = window.URL.createObjectURL(await (await fetch('data:{<MyCode>.GetMimeType(Path.GetExtension(filename))};base64,{Convert.ToBase64String(data)}')).blob()); "
		+ $"var link = document.createElement('a'); "
		+ $"link.href = data; "
		+ $"link.download='{filename}'; "
		+ $"link.click(); "
		+ $"setTimeout(function() {{ "
			+ $"window.URL.revokeObjectURL(data); "
		+ $"}}, 100); "
	+ $"}})();");
}
```

However, the downloaded file would not open in browser. Would you have any idea on what could be done?

**Reply by jeromelaban** [reply]
You may want to use the `FileSavePicker` instead which uses a similar technique.

**Reply by DierkDroth** [reply]
@jeromelaban hmm ... not sure I follow. 

I do have a few use cases of FileSavePicker in my app, but - obviously - this would be an additional dialog coming up - which is not what I needed, since I already have the file's data 'at hand'

Could you please elaborate on your suggestion (without an additional UI coming up)?

**Reply by jeromelaban** [reply]
Browsers generally don't make it easy open/launch something without the user interacting with it, you may still have to use that intermediate UI.

**Reply by DierkDroth** [reply]
Understand, thanks
